janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/AbstractIndexSelectionStrategy.java

```diff
@@ -0,0 +1,242 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.query.index;
+
+import javax.annotation.Nullable;
+import com.google.common.base.Preconditions;
+import org.janusgraph.core.*;
+import org.janusgraph.core.attribute.Cmp;
+import org.janusgraph.core.schema.JanusGraphSchemaType;
+import org.janusgraph.core.schema.SchemaStatus;
+import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.types.*;
+import org.janusgraph.graphdb.types.system.ImplicitKey;
+
+import java.util.*;
+import java.util.stream.StreamSupport;
+
+public abstract class AbstractIndexSelectionStrategy implements IndexSelectionStrategy {
+    private static final double EQUAL_CONDITION_SCORE = 4;
+    private static final double OTHER_CONDITION_SCORE = 1;
+    private static final double CARDINALITY_SINGE_SCORE = 1000;
+    private static final double CARDINALITY_OTHER_SCORE = 1000;
+
+    protected static final double ORDER_MATCH = 1;
+    protected static final double ALREADY_MATCHED_ADJUSTOR = -1.1;
```

Comment:
Missed on this. It is used only by `ApproximateIndexSelectionStrategy`, so I think it should be moved there

Reply:
I briefly thought about that, but I decided (for no reason) to have all constants in one place. I'm also fine with moving them to the strategy that uses them, if it is only a single one.

Reply:
I think it is better to move them into a concrete strategy while they are used only by that strategy